compliance_checker/cf/cf_1_6.py

```diff
@@ -1821,10 +1821,45 @@ def check_time_coordinate(self, ds):
             reasoning = None
             if not correct_units:
                 reasoning = ["%s does not have correct time units" % name]
-            result = Result(
-                BaseCheck.HIGH, correct_units, self.section_titles["4.4"], reasoning
+                result = Result(
+                    BaseCheck.HIGH, correct_units, self.section_titles["4.4"], reasoning
+                )
+                ret_val.append(result)
+                continue
+            # TODO: Is this a reasonable regex for extracting reference time
+            #       fields of interest?  Was unable to locate the grammar
+            #       or documentation for how reference times should be formed
```

Comment:
Wasn't aware of the granularity of this, thanks.

Reply:
Running the example modified from the test yields this result:

```
In [4]: Unit("months since 0-1-1 23:00:60").num2date(0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 Unit("months since 0-1-1 23:00:60").num2date(0)

File ~/.virtualenvs/cchecker_py3.9/lib/python3.9/site-packages/cf_units/__init__.py:2043, in Unit.num2date(self, time_value, only_use_cftime_datetimes, only_use_python_datetimes)
   1985 def num2date(
   1986     self,
   1987     time_value,
   1988     only_use_cftime_datetimes=True,
   1989     only_use_python_datetimes=False,
   1990 ):
   1991     """
   1992     Returns a datetime-like object calculated from the numeric time
   1993     value using the current calendar and the unit time reference.
   (...)
   2040 
   2041     """
-> 2043     return _num2date_to_nearest_second(
   2044         time_value,
   2045         self,
   2046         only_use_cftime_datetimes=only_use_cftime_datetimes,
   2047         only_use_python_datetimes=only_use_python_datetimes,
   2048     )

File ~/.virtualenvs/cchecker_py3.9/lib/python3.9/site-packages/cf_units/__init__.py:553, in _num2date_to_nearest_second(time_value, unit, only_use_cftime_datetimes, only_use_python_datetimes)
    542 time_values = time_values.ravel()
    544 # We account for the edge case where the time is in seconds and has a
    545 # half second: cftime.num2date() may produce a date that would round
    546 # down.
   (...)
    551 # arbitrarily. It is probably not possible to replicate that behaviour with
    552 # later versions, if one wished to do so for the sake of consistency.
--> 553 cftime_unit = unit.cftime_unit
    554 time_units = cftime_unit.split(" ")[0]
    555 has_half_seconds = np.logical_and(
    556     time_units == "seconds", time_values % 1.0 == 0.5
    557 )

File ~/.virtualenvs/cchecker_py3.9/lib/python3.9/site-packages/cf_units/__init__.py:1939, in Unit.cftime_unit(self)
   1934     interval = self.origin.split(" ")[0]
   1935     emsg = (
   1936         'Time units with interval of "months", "years" '
   1937         '(or singular of these) cannot be processed, got "{!s}".'
   1938     )
-> 1939     raise ValueError(emsg.format(interval))
   1941 #
   1942 # ensure to strip out non-parsable 'UTC' postfix, which
   1943 # is generated by UDUNITS-2 formatted output
   1944 #
   1945 return str(self).rstrip(" UTC")
```

So while there are three things wrong (months unit, year zero, seconds > 60 [WILLFIX in code per comment above]), only one error comes up.

While such a degenerately formed date is not common, it'd be good to nonetheless report all the errors.  Thoughts?

Reply:
That's a fair point. You'd also need to distinguish between the sec>=60 case, which is an error, while the other two should only be warnings.

You could use a simpler regex pattern to get just the "<units> since" part, and then use `cftime.datetime` (or `datetime.datetime`) to parse the reference time?

Reply:
cftime is fine with year 0 BTW.

```python
----> 1 cftime.datetime(0, 1, 1, 23, 00, 60)

File src/cftime/_cftime.pyx:1158, in cftime._cftime.datetime.__init__()

File src/cftime/_cftime.pyx:1767, in cftime._cftime.assert_valid_date()

ValueError: invalid second provided in cftime.datetime(0, 1, 1, 23, 0, 60, 0, calendar='standard', has_year_zero=True)
```

but

```python
In [6]: cftime.datetime(0, 1, 1, 23, 00, 59)
<ipython-input-6-2411681357cf>:1: CFWarning: year=0 was specified - this date/calendar/year zero convention is not supported by CF
  cftime.datetime(0, 1, 1, 23, 00, 59)
<ipython-input-6-2411681357cf>:1: CFWarning: this date/calendar/year zero convention is not supported by CF
  cftime.datetime(0, 1, 1, 23, 00, 59)
Out[6]: cftime.datetime(0, 1, 1, 23, 0, 59, 0, calendar='standard', has_year_zero=True)
```

We got away with just warning and maybe that is even valid for some other calendars (untested) :-/

Reply:
BTW, for a units test I would do:

```python
Unit("months since 0-1-1 23:00:60").cftime_unit
```

which yields:

```
ValueError: Time units with interval of "months", "years" (or singular of these) cannot be processed, got "months".
```

Reply:
OK, like I though, the year zero is a warning b/c we can only error out if we know the calendar, which we don't in your test above. See this part from the cftime 1.5.1 README:

    10/1/2021: Version 1.5.1 released. Changed default behavior of proleptic_gregorian to has_year_zero=T
    (since it is allowed in ISO-8601 and CF does not specify the year zero convention for this calendar).
    Raise warning message when trying to create a calendar that is not supported by CF version 1.9
    (no years < 1 allowed for 'standard'/'gregorian' or 'julian' calendars).
    Added support for "common_year" and "common_years" units for "noleap" and "365_day" calendars.

xref.: https://github.com/Unidata/cftime/blob/7a0dc391a95a5df4b253ae0271bc3d11978f869b/README.md

Reply:
Ack, still need to handle climatology year 0 for non-standard calendars.  Also looks like there's a disagreement with conformance doc and the conventions doc, the former gives section 4.4, and the latter gives 4.4.1 and also takes priority the former and is more nuanced.  Will notify CF conventions team on this.